File: Greedy/JobSequencing/JobSequencing.js
class Job {
  constructor(id, deadline, profit) {
    this.id = id;
    this.deadline = deadline;
    this.profit = profit;
  }
}

function method1(arr) {
  let max = 0,i,j,count=0;
  arr.sort((a,b)=>{
    max = Math.max(max,Math.max(a.deadline,b.deadline));
    return b.profit-a.profit;
  });
  console.log('arr is ', arr);
  //console.log('max is ', max);
  let tab = new Array(max+1);
  for(i=0;i<arr.length;i++){
    if(count==max){
      break;
    }
    if(tab[arr[i].deadline]){
      j=arr[i].deadline-1;
      while(tab[j]){
        j--;
      }
      if(j>0){
        tab[j]=arr[i].profit;
        count--;
      }
    }else{
      tab[arr[i].deadline]=arr[i].profit;
      count--;
    }
  }
  let sum=0;
  //console.log(tab);
  for(i=1;i<tab.length;i++)
    if(tab[i])
      sum+=tab[i];
  return sum;
}


let arr = [
  new Job(1,2,100),
  new Job(2,1,19),
  new Job(3,2,27),
  new Job(4,1,25),
  new Job(5,1,15)
];

console.log(method1(arr));

File: Greedy/JobSequencing/JobSequencing.go
package main

import (
	"fmt"
	"sort"
)

type job struct {
	id       int
	deadline int
	profit   int
}

func (arr jobList) Len() int {
	return len(arr)
}

func (arr jobList) Less(i, j int) bool {
	return arr[i].profit > arr[j].profit
}

func (arr jobList) Swap(i, j int) {
	arr[i], arr[j] = arr[j], arr[i]
}

func method1(arr jobList) int {
	sort.Sort(arr)
	n := len(arr)
	max := 0
	var i int
	for i = 0; i < n; i++ {
		if arr[i].deadline > max {
			max = arr[i].deadline
		}
	}
	visited := make([]bool, max)
	sum := 0
	count := 0

	for i := 0; i < n; i++ {
		if count == max {
			return sum
		}
		if visited[arr[i].deadline-1] == true {
			for j := arr[i].deadline - 2; j >= 0; j-- {
				if visited[j] == false {
					sum += arr[i].profit
					visited[j] = true
					count++
					break
				}
			}
		} else {
			sum += arr[i].profit
			visited[arr[i].deadline-1] = true
			count++
		}
	}
	return sum
}

type jobList []job

func main() {
	jobList := []job{
		job{1, 4, 20},
		job{2, 1, 10},
		job{3, 1, 40},
		job{4, 1, 30},
	}
	fmt.Println(method1(jobList))
	fmt.Println(jobList)
}
